File: cli/metalk.py
"""MeTalk — caveman-style token compression for CRUMB.

Layers:
  Layer 1 (lossless):    dictionary word→abbreviation swaps
  Layer 2 (lossy):       strip articles/filler, rewrite verbose phrases
  Layer 3 (aggressive):  + sentence condensing
  Layer 4 (skeleton):    + interior vowel stripping (cli/vowelstrip)
  Layer 5 (adaptive):    + embedding-aware vowel stripping (requires
                         the [embeddings] extra; falls back to L4)

Usage:
    from cli.metalk import encode, decode, compression_stats
    slim = encode(crumb_text, level=2)
    full = decode(slim)  # reverses Layer 1 only
"""

import re

# ── Layer 1: Dictionary substitutions (lossless, reversible) ─────────

STRUCTURAL = {
    "BEGIN CRUMB": "BC",
    "END CRUMB": "EC",
}

SECTION_MAP = {
    "goal": "g",
    "context": "cx",
    "constraints": "ct",
    "consolidated": "cs",
    "project": "pj",
    "modules": "md",
    "entries": "en",
    "tasks": "tk",
    "identity": "id",
    "permissions": "pm",
    "actions": "ac",
    "verdict": "vd",
    "notes": "nt",
    "dream": "dm",
    "raw": "rw",
    "archived": "ar",
    "summary": "sm",
    "findings": "fd",
    "recommendations": "rc",
}

HEADER_KEY_MAP = {
    "kind": "k",
    "source": "src",
    "title": "t",
    "project": "pj",
    "max_index_tokens": "mit",
    "max_total_tokens": "mtt",
    "dream_pass": "dp",
    "dream_sessions": "ds",
    "agent_name": "an",
    "agent_framework": "af",
    "status": "st",
    "issued": "iss",
    "expires": "exp",
    "metalk": "mt",
}

ABBREV = {
    # Long tech terms → short forms (sorted longest-first at runtime)
    "authentication": "auth",
    "authorization": "authz",
    "configuration": "cfg",
    "implementation": "impl",
    "infrastructure": "infra",
    "documentation": "docs",
    "specification": "spec",
    "asynchronous": "async",
    "synchronous": "sync",
    "architecture": "arch",
    "notification": "notif",
    "application": "app",
    "environment": "env",
    "development": "dev",
    "performance": "perf",
    "distributed": "distrib",
    "dependencies": "deps",
    "requirements": "reqs",
    "deployment": "deploy",
    "integration": "integ",
    "certificate": "cert",
    "credential": "cred",
    "continuous": "cont",
    "expression": "expr",
    "connection": "conn",
    "transaction": "txn",
    "dependency": "dep",
    "repository": "repo",
    "kubernetes": "k8s",
    "middleware": "mw",
    "production": "prod",
    "components": "cmps",
    "properties": "props",
    "parameters": "params",
    "attributes": "attrs",
    "permission": "perm",
    "validation": "val",
    "definition": "defn",
    "execution": "exec",
    "component": "cmp",
    "container": "ctnr",
    "publisher": "pub",
    "subscriber": "sub",
    "variables": "vars",
    "operation": "op",
    "reference": "ref",
    "parameter": "param",
    "attribute": "attr",
    "directory": "dir",
    "interface": "ifc",
    "exception": "exc",
    "postgresql": "pg",
    "typescript": "ts",
    "javascript": "js",
    "template": "tmpl",
    "property": "prop",
    "function": "fn",
    "database": "db",
    "response": "resp",
    "variable": "var",
    "callback": "cb",
    "message": "msg",
    "request": "rq",
    "information": "info",
}

# Build reverse maps
_REV_STRUCTURAL = {v: k for k, v in STRUCTURAL.items()}
_REV_SECTION = {v: k for k, v in SECTION_MAP.items()}
_REV_HEADER_KEY = {v: k for k, v in HEADER_KEY_MAP.items()}
_REV_ABBREV = {v: k for k, v in ABBREV.items()}

# Pre-sort by length descending for longest-match-first
_ABBREV_SORTED = sorted(ABBREV.items(), key=lambda x: len(x[0]), reverse=True)
_REV_ABBREV_SORTED = sorted(_REV_ABBREV.items(), key=lambda x: len(x[0]), reverse=True)


# ── Layer 2: Grammar stripping (lossy) ──────────────────────────────

STRIP_WORDS = frozenset({
    "the", "a", "an", "that", "which", "just", "very", "really",
    "basically", "actually", "currently", "essentially", "simply",
    "furthermore", "moreover", "however", "therefore", "consequently",
    "additionally", "specifically", "particularly", "approximately",
})

PHRASE_REWRITES = {
    "in order to": "to",
    "make sure to": "ensure",
    "as well as": "and",
    "due to the fact that": "because",
    "at this point in time": "now",
    "in the event that": "if",
    "it is necessary to": "must",
    "there is a need to": "must",
    "is able to": "can",
    "be able to": "can",
    "needs to be": "must be",
    "do not": "don't",
    "does not": "doesn't",
    "should not": "shouldn't",
    "is not": "isn't",
    "can not": "can't",
    "cannot": "can't",
    "has been": "was",
    "have been": "were",
    "will need to": "must",
    "you should": "",
    "please note that": "",
    "it is important to note that": "",
    "it should be noted that": "",
}

# Sort phrase rewrites longest-first
_PHRASE_SORTED = sorted(PHRASE_REWRITES.items(), key=lambda x: len(x[0]), reverse=True)


# ── Core encode/decode ──────────────────────────────────────────────

def _estimate_tokens(text: str) -> int:
    return max(1, len(text) // 4)


def _apply_dict_sub(text: str, mapping: list, body_only_start: int = 0) -> str:
    """Apply whole-word substitutions from a sorted mapping list."""
    for long_form, short_form in mapping:
        pattern = re.compile(r'\b' + re.escape(long_form) + r'\b', re.IGNORECASE)
        def _replace(m):
            orig = m.group(0)
            # Preserve capitalization
            if orig[0].isupper() and short_form[0].islower():
                return short_form[0].upper() + short_form[1:]
            return short_form
        text = pattern.sub(_replace, text)
    return text


def _reverse_dict_sub(text: str, mapping: list) -> str:
    """Reverse whole-word substitutions."""
    for short_form, long_form in mapping:
        pattern = re.compile(r'\b' + re.escape(short_form) + r'\b')
        def _replace(m, lf=long_form):
            orig = m.group(0)
            if orig[0].isupper() and lf[0].islower():
                return lf[0].upper() + lf[1:]
            return lf
        text = pattern.sub(_replace, text)
    return text


def _strip_grammar(text: str) -> str:
    """Remove articles, filler words, and rewrite verbose phrases."""
    # Apply phrase rewrites first (longest-first)
    for phrase, replacement in _PHRASE_SORTED:
        pattern = re.compile(re.escape(phrase), re.IGNORECASE)
        text = pattern.sub(replacement, text)

    # Strip filler words (whole-word, preserve line structure)
    lines = text.split('\n')
    result = []
    for line in lines:
        words = line.split()
        filtered = []
        for w in words:
            # Check stripped version (without punctuation) against strip list
            clean = w.strip('.,;:!?()[]"\'').lower()
            if clean in STRIP_WORDS:
                continue
            filtered.append(w)
        new_line = ' '.join(filtered)
        # Clean up double spaces
        new_line = re.sub(r'  +', ' ', new_line).strip()
        # Fix orphaned leading punctuation
        new_line = re.sub(r'^- +', '- ', new_line)
        result.append(new_line)
    return '\n'.join(result)


def _condense_aggressive(text: str) -> str:
    """Level 3: aggressive sentence condensing."""
    lines = text.split('\n')
    result = []
    for line in lines:
        stripped = line.strip()
        # Remove trailing periods on bullet points
        if stripped.startswith('-') and stripped.endswith('.'):
            line = line.rstrip()
            if line.endswith('.'):
                line = line[:-1]
        # Remove empty bullet points
        if stripped == '-' or stripped == '- ':
            continue
        result.append(line)

    # Collapse multiple blank lines
    text = '\n'.join(result)
    text = re.sub(r'\n{3,}', '\n\n', text)
    return text


def encode(text: str, level: int = 2, *,
           vowel_min_length: int = 4,
           adaptive_threshold: float = 0.85) -> str:
    """Compress crumb text using MeTalk.

    Args:
        text: Full crumb text (BEGIN CRUMB ... END CRUMB)
        level: 1=dict only (lossless), 2=dict+grammar (lossy),
               3=aggressive, 4=skeleton (vowel-strip),
               5=adaptive (embedding-aware vowel-strip; needs the
               [embeddings] extra, falls back to 4 if unavailable).
        vowel_min_length: Min word length eligible for vowel stripping
                          at levels 4-5. Default 4.
        adaptive_threshold: Cosine similarity floor for level 5 — a
                            line keeps its stripped form only if drift
                            stays under this. Default 0.85.

    Returns:
        MeTalk-encoded crumb text with mt=N header injected.
    """
    lines = text.strip().split('\n')

    # Find the separator
    sep_idx = None
    for i, line in enumerate(lines):
        if line.strip() == '---':
            sep_idx = i
            break

    if sep_idx is None:
        # Not a valid crumb structure, just apply body transforms
        return text

    # Phase A: Structural markers
    if lines[0].strip() == 'BEGIN CRUMB':
        lines[0] = 'BC'
    if lines[-1].strip() == 'END CRUMB':
        lines[-1] = 'EC'

    # Phase B: Header keys (before ---)
    header_lines = []
    mt_injected = False
    for i in range(1, sep_idx):
        line = lines[i]
        if '=' in line:
            key, _, val = line.partition('=')
            key_stripped = key.strip()
            if key_stripped in HEADER_KEY_MAP:
                line = f"{HEADER_KEY_MAP[key_stripped]}={val}"
        header_lines.append(line)

    # Inject mt=N header
    header_lines.append(f"mt={level}")

    # Phase C: Section headers and body (after ---)
    body_lines = []
    for i in range(sep_idx + 1, len(lines) - 1):  # exclude EC
        line = lines[i]
        stripped = line.strip()

        # Section headers
        if stripped.startswith('[') and stripped.endswith(']'):
            section_name = stripped[1:-1].strip().lower()
            if section_name in SECTION_MAP:
                indent = line[:len(line) - len(line.lstrip())]
                line = f"{indent}[{SECTION_MAP[section_name]}]"
            body_lines.append(line)
            continue

        # Body text: apply dictionary substitutions
        line = _apply_dict_sub(line, _ABBREV_SORTED)

        # Layer 2: grammar stripping
        if level >= 2:
            line = _strip_grammar(line)

        body_lines.append(line)

    # Layer 3: aggressive condensing
    if level >= 3:
        body_text = '\n'.join(body_lines)
        body_text = _condense_aggressive(body_text)
        body_lines = body_text.split('\n')

    # Reassemble
    result_lines = [lines[0]]  # BC
    result_lines.extend(header_lines)
    result_lines.append('---')
    result_lines.extend(body_lines)
    result_lines.append(lines[-1])  # EC

    assembled = '\n'.join(result_lines) + '\n'

    # Layers 4 and 5: vowel-strip pass over the body. Done last so that
    # dictionary substitutions (Layer 1) have already canonicalized the
    # most ambiguous tech terms before vowels are removed.
    if level >= 4:
        from cli.vowelstrip import encode_crumb as _vs_encode, adaptive_strip_text, strip_line
        if level >= 5:
            transform = lambda line: adaptive_strip_text(
                line, threshold=adaptive_threshold, min_length=vowel_min_length
            )
        else:
            transform = lambda line: strip_line(line, vowel_min_length)
        assembled = _vs_encode(assembled, min_length=vowel_min_length, transform=transform)

    return assembled


def decode(text: str) -> str:
    """Decode MeTalk-encoded crumb. Reverses Layer 1 (dictionary) only.

    Layers 2-5 are lossy and cannot be fully reversed. Vowel-stripped
    words from L4/L5 pass through unchanged (their consonant skeletons
    don't match the dictionary). The `vs=` header is stripped on decode.
    """
    lines = text.strip().split('\n')

    # Check for mt= header
    mt_level = None
    for line in lines:
        stripped = line.strip()
        if stripped.startswith('mt='):
            try:
                mt_level = int(stripped.split('=', 1)[1])
            except ValueError:
                pass
            break

    if mt_level is None:
        return text  # Not MeTalk-encoded, passthrough

    # Phase A: Structural markers
    if lines[0].strip() == 'BC':
        lines[0] = 'BEGIN CRUMB'
    if lines[-1].strip() == 'EC':
        lines[-1] = 'END CRUMB'

    # Find separator
    sep_idx = None
    for i, line in enumerate(lines):
        if line.strip() == '---':
            sep_idx = i
            break

    if sep_idx is None:
        return '\n'.join(lines) + '\n'

    # Phase B: Header keys
    header_lines = []
    for i in range(1, sep_idx):
        line = lines[i]
        stripped = line.strip()
        # Skip mt= and vs= MeTalk-internal headers (remove them)
        if stripped.startswith('mt=') or stripped.startswith('vs='):
            continue
        if '=' in line:
            key, _, val = line.partition('=')
            key_stripped = key.strip()
            if key_stripped in _REV_HEADER_KEY:
                line = f"{_REV_HEADER_KEY[key_stripped]}={val}"
        header_lines.append(line)

    # Phase C: Section headers and body
    body_lines = []
    for i in range(sep_idx + 1, len(lines) - 1):
        line = lines[i]
        stripped = line.strip()

        # Section headers
        if stripped.startswith('[') and stripped.endswith(']'):
            section_name = stripped[1:-1].strip().lower()
            if section_name in _REV_SECTION:
                indent = line[:len(line) - len(line.lstrip())]
                line = f"{indent}[{_REV_SECTION[section_name]}]"
            body_lines.append(line)
            continue

        # Reverse dictionary substitutions
        line = _reverse_dict_sub(line, _REV_ABBREV_SORTED)
        body_lines.append(line)

    # Reassemble
    result_lines = [lines[0]]  # BEGIN CRUMB
    result_lines.extend(header_lines)
    result_lines.append('---')
    result_lines.extend(body_lines)
    result_lines.append(lines[-1])  # END CRUMB

    return '\n'.join(result_lines) + '\n'


def compression_stats(original: str, encoded: str) -> dict:
    """Return compression statistics."""
    orig_tokens = _estimate_tokens(original)
    enc_tokens = _estimate_tokens(encoded)
    saved = orig_tokens - enc_tokens
    pct = (saved / orig_tokens * 100) if orig_tokens > 0 else 0
    ratio = orig_tokens / enc_tokens if enc_tokens > 0 else 1
    return {
        "original_tokens": orig_tokens,
        "encoded_tokens": enc_tokens,
        "saved_tokens": saved,
        "pct_saved": round(pct, 1),
        "ratio": round(ratio, 2),
        "original_chars": len(original),
        "encoded_chars": len(encoded),
    }

File: browser-extension/metalk.js
/**
 * Pure-JS port of MeTalk + vowel-strip (cli/metalk.py + cli/vowelstrip.py).
 *
 * The dict/set tables live in metalk-data.json (exported from the Python
 * source of truth by scripts/export_metalk_data.py). This module loads the
 * JSON once on first use and exposes an API that matches the Python surface.
 *
 * Exports (both browser `window.Metalk` and ES-module form):
 *   encode(text, level, opts?)     — full pipeline, levels 1-5 (5 falls back to 4)
 *   decode(text)                   — reverse Layer 1 dict subs; strip mt=/vs= headers
 *   stripText(text, minLen)        — vowel-strip plain prose
 *   stripLine(line, minLen)        — vowel-strip one line
 *   compressionStats(orig, enc)    — {original_tokens, encoded_tokens, …}
 *   load(url?)                     — manually preload the data JSON
 *
 * Adaptive (level 5) has no in-browser embedding model by default, so it
 * falls back to level 4. The browser extension can override by passing a
 * per-line predicate via opts.keepStrip(line, candidate).
 */
(function (root, factory) {
  "use strict";
  if (typeof module === "object" && module.exports) {
    module.exports = factory();
  } else {
    root.Metalk = factory();
  }
})(typeof self !== "undefined" ? self : this, function () {
  "use strict";

  var DATA = null;
  var DATA_URL_DEFAULT = "metalk-data.json";

  // ── Data loading ────────────────────────────────────────────

  function load(url) {
    var target = url || DATA_URL_DEFAULT;
    return fetch(target).then(function (r) {
      if (!r.ok) throw new Error("metalk-data.json fetch failed: " + r.status);
      return r.json();
    }).then(function (data) {
      setData(data);
      return data;
    });
  }

  function setData(data) {
    DATA = data;
    // Pre-build sorted (longest-first) lists for regex building.
    DATA._abbrevSorted = Object.keys(data.abbrev)
      .sort(function (a, b) { return b.length - a.length; })
      .map(function (k) { return [k, data.abbrev[k]]; });
    DATA._phrasesSorted = Object.keys(data.phrase_rewrites)
      .sort(function (a, b) { return b.length - a.length; })
      .map(function (k) { return [k, data.phrase_rewrites[k]]; });
    DATA._abbrevReverseSorted = Object.keys(data.abbrev)
      .map(function (k) { return [data.abbrev[k], k]; })
      .sort(function (a, b) { return b[0].length - a[0].length; });
    DATA._stripSet = new Set(data.strip_words.map(function (w) { return w.toLowerCase(); }));
    DATA._protectedSet = new Set(data.protected_words.map(function (w) { return w.toLowerCase(); }));
    DATA._punctSet = new Set(data.sentence_punct);
    DATA._opaqueSet = new Set(data.opaque_chars);
    DATA._vowelSet = new Set(data.vowels);
    return DATA;
  }

  function ensureData() {
    if (!DATA) throw new Error("Metalk data not loaded. Call Metalk.load() first.");
  }

  // ── Helpers ─────────────────────────────────────────────────

  function escapeRegExp(s) {
    return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
  }

  function estimateTokens(text) {
    return Math.max(1, Math.floor(text.length / 4));
  }

  function matchCase(target, sourceFirst) {
    // Preserve capitalization of first character from source.
    if (!target) return target;
    if (sourceFirst && sourceFirst[0].toUpperCase() === sourceFirst[0] && target[0].toLowerCase() === target[0]) {
      return target[0].toUpperCase() + target.slice(1);
    }
    return target;
  }

  // ── Layer 1: dictionary substitution ─────────────────────────

  function applyDict(text, sortedPairs) {
    var out = text;
    sortedPairs.forEach(function (pair) {
      var longForm = pair[0], shortForm = pair[1];
      var re = new RegExp("\\b" + escapeRegExp(longForm) + "\\b", "gi");
      out = out.replace(re, function (m) { return matchCase(shortForm, m); });
    });
    return out;
  }

  function reverseDict(text, sortedPairs) {
    var out = text;
    sortedPairs.forEach(function (pair) {
      var shortForm = pair[0], longForm = pair[1];
      var re = new RegExp("\\b" + escapeRegExp(shortForm) + "\\b", "g");
      out = out.replace(re, function (m) { return matchCase(longForm, m); });
    });
    return out;
  }

  // ── Layer 2: grammar stripping ───────────────────────────────

  function stripGrammar(text) {
    // Phrase rewrites first (longest-first).
    DATA._phrasesSorted.forEach(function (pair) {
      var re = new RegExp(escapeRegExp(pair[0]), "gi");
      text = text.replace(re, pair[1]);
    });
    // Filler words per line, preserving structure.
    return text.split("\n").map(function (line) {
      var words = line.split(/\s+/);
      var kept = [];
      words.forEach(function (w) {
        var clean = w.replace(/^[.,;:!?()[\]"']+|[.,;:!?()[\]"']+$/g, "").toLowerCase();
        if (DATA._stripSet.has(clean)) return;
        kept.push(w);
      });
      return kept.join(" ").replace(/  +/g, " ").replace(/^- +/, "- ").replace(/^\s+|\s+$/g, "");
    }).join("\n");
  }

  // ── Layer 3: aggressive condense ─────────────────────────────

  function condenseAggressive(text) {
    var out = text.split("\n").filter(function (line) {
      var s = line.trim();
      return s !== "-" && s !== "- ";
    }).map(function (line) {
      var s = line.trim();
      if (s.startsWith("-") && s.endsWith(".")) return line.replace(/\.$/, "");
      return line;
    }).join("\n");
    return out.replace(/\n{3,}/g, "\n\n");
  }

  // ── Layer 4/5: vowel strip ──────────────────────────────────

  function shouldStrip(word, minLen) {
    if (word.length < minLen) return false;
    if (word === word.toUpperCase()) return false; // acronym
    if (DATA._protectedSet.has(word.toLowerCase())) return false;
    return true;
  }

  function stripWord(word, minLen) {
    if (minLen === undefined) minLen = 4;
    if (!/^[A-Za-z]+$/.test(word)) return word;
    if (!shouldStrip(word, minLen)) return word;

    var first = word[0];
    var body = word.slice(1);
    var keepS = body.length > 1 && (body.endsWith("s") || body.endsWith("S"));
    var core = keepS ? body.slice(0, -1) : body;
    var stripped = "";
    for (var i = 0; i < core.length; i++) {
      if (!DATA._vowelSet.has(core[i])) stripped += core[i];
    }
    var result = first + stripped + (keepS ? body[body.length - 1] : "");
    if (result.length <= 1 && word.length >= minLen) return word.slice(0, 2);
    return result;
  }

  function isOpaqueToken(tok) {
    for (var i = 0; i < tok.length; i++) {
      if (DATA._opaqueSet.has(tok[i])) return true;
    }
    return false;
  }

  function stripToken(tok, minLen) {
    var leading = "";
    while (tok.length && DATA._punctSet.has(tok[0])) { leading += tok[0]; tok = tok.slice(1); }
    var trailingRev = "";
    while (tok.length && DATA._punctSet.has(tok[tok.length - 1])) {
      trailingRev = tok[tok.length - 1] + trailingRev;
      tok = tok.slice(0, -1);
    }
    if (!tok) return leading + trailingRev;
    var core = isOpaqueToken(tok)
      ? tok
      : tok.replace(/[A-Za-z]+/g, function (m) { return stripWord(m, minLen); });
    return leading + core + trailingRev;
  }

  function stripLine(line, minLen) {
    if (minLen === undefined) minLen = 4;
    ensureData();
    return line.split(/(\s+)/).map(function (part) {
      if (!part || /^\s+$/.test(part)) return part;
      return stripToken(part, minLen);
    }).join("");
  }

  function stripText(text, minLen) {
    return text.split("\n").map(function (l) { return stripLine(l, minLen); }).join("\n");
  }

  // ── CRUMB-aware vowel strip ──────────────────────────────────

  var SECTION_RE = /^\s*\[[^\]]+\]\s*$/;
  var FENCE_RE = /^\s*```/;
  var TYPED_RE = /^\s*@type\s*:\s*(code|diff|json|yaml)\b/i;

  function stripCrumb(text, opts) {
    opts = opts || {};
    var minLen = opts.min_length || 4;
    var transform = opts.transform || function (l) { return stripLine(l, minLen); };

    var lines = text.replace(/\n+$/, "").split("\n");
    var sepIdx = -1;
    for (var i = 0; i < lines.length; i++) {
      if (lines[i].trim() === "---") { sepIdx = i; break; }
    }
    if (sepIdx === -1) return text;

    var out = lines.slice(0, sepIdx + 1);
    var inFence = false;
    var inTypedCode = false;

    for (var j = sepIdx + 1; j < lines.length; j++) {
      var line = lines[j];
      if (SECTION_RE.test(line)) { inTypedCode = false; out.push(line); continue; }
      if (FENCE_RE.test(line)) { inFence = !inFence; out.push(line); continue; }
      var m = TYPED_RE.exec(line);
      if (m) {
        var kind = m[1].toLowerCase();
        inTypedCode = ["code", "diff", "json", "yaml"].indexOf(kind) >= 0;
        out.push(line); continue;
      }
      if (line.trim() === "END CRUMB" || line.trim() === "EC") { out.push(line); continue; }
      if (inFence || inTypedCode) { out.push(line); continue; }
      out.push(transform(line));
    }
    return injectVsHeader(out.join("\n") + "\n", minLen);
  }

  function injectVsHeader(encoded, minLen) {
    var lines = encoded.split("\n");
    var sepIdx = -1;
    for (var i = 0; i < lines.length; i++) if (lines[i].trim() === "---") { sepIdx = i; break; }
    if (sepIdx === -1) return encoded;
    for (var k = 1; k < sepIdx; k++) {
      if (lines[k].trim().startsWith("vs=")) { lines[k] = "vs=" + minLen; return lines.join("\n"); }
    }
    lines.splice(sepIdx, 0, "vs=" + minLen);
    return lines.join("\n");
  }

  // ── Encode / decode pipelines ───────────────────────────────

  function encodePlain(text, level, opts) {
    // Pure body-transform pipeline for arbitrary prose. Mirrors
    // cli/metalk.py::encode_plain — never treats `[bracket]` lines as
    // section headers, so user headings like `[goal]` or `[context]` are
    // preserved verbatim.
    ensureData();
    opts = opts || {};
    level = level || 2;
    var vml = opts.vowel_min_length || 4;
    var result = applyDict(text, DATA._abbrevSorted);
    if (level >= 2) result = stripGrammar(result);
    if (level >= 3) result = condenseAggressive(result);
    if (level >= 4) result = stripText(result, vml);
    return result;
  }

  function encode(text, level, opts) {
    ensureData();
    opts = opts || {};
    level = level || 2;
    var vml = opts.vowel_min_length || 4;

    var lines = text.trim().split("\n");
    var sepIdx = lines.indexOf("---");
    if (sepIdx === -1) {
      // Not a structured crumb — run the plain pipeline.
      return encodePlain(text, level, opts);
    }

    if (lines[0].trim() === "BEGIN CRUMB") lines[0] = "BC";
    if (lines[lines.length - 1].trim() === "END CRUMB") lines[lines.length - 1] = "EC";

    var headerLines = [];
    for (var i = 1; i < sepIdx; i++) {
      var line = lines[i];
      if (line.includes("=")) {
        var idx = line.indexOf("=");
        var key = line.slice(0, idx).trim();
        var val = line.slice(idx + 1);
        if (DATA.header_key_map[key]) line = DATA.header_key_map[key] + "=" + val;
      }
      headerLines.push(line);
    }
    headerLines.push("mt=" + level);

    var bodyLines = [];
    for (var j = sepIdx + 1; j < lines.length - 1; j++) {
      var bl = lines[j];
      var stripped = bl.trim();
      if (stripped.startsWith("[") && stripped.endsWith("]")) {
        var name = stripped.slice(1, -1).trim().toLowerCase();
        if (DATA.section_map[name]) {
          var indent = bl.match(/^\s*/)[0];
          bl = indent + "[" + DATA.section_map[name] + "]";
        }
        bodyLines.push(bl);
        continue;
      }
      bl = applyDict(bl, DATA._abbrevSorted);
      if (level >= 2) bl = stripGrammar(bl);
      bodyLines.push(bl);
    }

    if (level >= 3) {
      var bodyText = bodyLines.join("\n");
      bodyText = condenseAggressive(bodyText);
      bodyLines = bodyText.split("\n");
    }

    var result2 = [lines[0]].concat(headerLines, ["---"], bodyLines, [lines[lines.length - 1]]).join("\n") + "\n";

    if (level >= 4) {
      var transform = function (ln) { return stripLine(ln, vml); };
      if (level >= 5 && typeof opts.keepStrip === "function") {
        transform = function (ln) {
          var cand = stripLine(ln, vml);
          return opts.keepStrip(ln, cand) ? cand : ln;
        };
      }
      result2 = stripCrumb(result2, { min_length: vml, transform: transform });
    }
    return result2;
  }

  function decode(text) {
    ensureData();
    var lines = text.trim().split("\n");
    var hasMt = lines.some(function (l) { return l.trim().startsWith("mt="); });
    if (!hasMt) return text;

    if (lines[0].trim() === "BC") lines[0] = "BEGIN CRUMB";
    if (lines[lines.length - 1].trim() === "EC") lines[lines.length - 1] = "END CRUMB";

    var sepIdx = lines.indexOf("---");
    if (sepIdx === -1) return lines.join("\n") + "\n";

    var headerLines = [];
    for (var i = 1; i < sepIdx; i++) {
      var line = lines[i];
      var stripped = line.trim();
      if (stripped.startsWith("mt=") || stripped.startsWith("vs=")) continue;
      if (line.includes("=")) {
        var idx = line.indexOf("=");
        var key = line.slice(0, idx).trim();
        var val = line.slice(idx + 1);
        for (var k in DATA.header_key_map) {
          if (DATA.header_key_map[k] === key) { line = k + "=" + val; break; }
        }
      }
      headerLines.push(line);
    }

    var bodyLines = [];
    for (var j = sepIdx + 1; j < lines.length - 1; j++) {
      var bl = lines[j];
      var strip = bl.trim();
      if (strip.startsWith("[") && strip.endsWith("]")) {
        var name = strip.slice(1, -1).trim().toLowerCase();
        for (var sk in DATA.section_map) {
          if (DATA.section_map[sk] === name) {
            var indent = bl.match(/^\s*/)[0];
            bl = indent + "[" + sk + "]";
            break;
          }
        }
        bodyLines.push(bl);
        continue;
      }
      bl = reverseDict(bl, DATA._abbrevReverseSorted);
      bodyLines.push(bl);
    }

    return [lines[0]].concat(headerLines, ["---"], bodyLines, [lines[lines.length - 1]]).join("\n") + "\n";
  }

  function compressionStats(original, encoded) {
    var ot = estimateTokens(original), et = estimateTokens(encoded);
    var saved = ot - et;
    return {
      original_tokens: ot,
      encoded_tokens: et,
      saved_tokens: saved,
      pct_saved: Math.round((saved / ot) * 1000) / 10,
      ratio: Math.round((ot / Math.max(et, 1)) * 100) / 100,
      original_chars: original.length,
      encoded_chars: encoded.length
    };
  }

  return {
    load: load,
    setData: setData,
    encode: encode,
    encodePlain: encodePlain,
    decode: decode,
    stripText: stripText,
    stripLine: stripLine,
    stripWord: stripWord,
    compressionStats: compressionStats,
    estimateTokens: estimateTokens
  };
});
